src/main.rs

```diff
@@ -18,50 +20,88 @@ mod crates_cache;
 mod publishers;
 mod subcommands;
 
-fn main() {
-    let mut args = std::env::args_os();
-    let _ = args.by_ref().next();
+#[derive(Debug)]
+struct Args {
+    help: bool,
+    command: String,
+    cache_max_age: Duration,
+    metadata_args: Vec<String>,
+}
 
-    while let Some(arg) = args.next() {
-        match arg.to_str() {
-            None => bail_bad_arg(arg),
-            Some("supply-chain") => (), // first arg when run as `cargo supply-chain`
-            Some("authors") => return subcommands::authors(args),
-            Some("publishers") => return subcommands::publishers(args),
-            Some("crates") => return subcommands::crates(args),
-            Some("update") => return subcommands::update(args),
-            Some(arg) if arg.starts_with("--") => bail_unknown_option(arg),
-            Some(arg) if arg.starts_with('-') => bail_unknown_short_option(arg),
-            Some(arg) => bail_unknown_command(arg),
+fn main() {
+    match args_parser() {
+        Ok(args) => match handle_args(args) {
+            Ok(_) => {}
+            Err(e) => {
+                eprintln!("Error: {}", e);
+            }
+        },
+        Err(e) => {
+            eprintln!("Error: {}", e);
+            eprint_help();
         }
     }
-
-    // No command selected.
-    bail_no_command();
 }
 
-fn bail_unknown_option(arg: &str) -> ! {
-    eprintln!("Unknown option: {}", std::path::Path::new(&arg).display());
-    eprint_help();
-    std::process::exit(1);
+fn handle_args(args: Args) -> Result<(), std::io::Error> {
+    if args.help {
+        eprint_help();
+    } else if args.command == "authors" {
+        subcommands::authors(args.metadata_args)
+    } else if args.command == "publishers" {
+        subcommands::publishers(args.metadata_args, args.cache_max_age)?;
+    } else if args.command == "crates" {
+        subcommands::crates(args.metadata_args, args.cache_max_age)?;
+    } else if args.command == "update" {
+        subcommands::update(args.cache_max_age)
```

Comment:
No changes needed, just FYI: this is usually handled with a `match` statement rather an an `else if` chain. But I believe the `else if` chain is OK in this particular case.

Reply:
ah, do you mean like
```
match args.command {
      "crates" => subcommands::crates...,
      "update" => ...,
```?

Reply:
Yes, exactly.